Add a Misc Section to Views Integration (#2677)
moderntribe/the-events-calendar: tests/views_integration/README.md

@todo @paulmskim

### Other Testing

> Besides all the tests mentioned above due to the multiple moving parts of the software is important to make sure we test unwanted behavior to make sure they are not being "fixed" in a future bug report.

Reviewer: ```suggestion
Besides all the tests mentioned above, due to the multiple moving parts of the software, its important to make sure we cover the current behavior with tests to ensure they are not broken in the next iterations.
```

> The test below is dealing with the multiple possibilities of the template bootstraping, and how it is based on an admin setting, which can be changed in a lot of unpredictable ways, so we bake into our integration testing a way to make sure even if those bad values are pulled from the database we still continue to behave as expoected.

Reviewer: ```suggestion
In the test below we cover the class bootstrapping the templates with tests to ensure it handles all the possible admin setting combinations, good and bad alike, in a predictable way.
Any change of the code causing a test failure would signal a change in the class behavior we want to avoid.
```

```php
<?php
namespace Tribe\Events\Views\V2;

class TemplateBootstrapTest extends \Codeception\TestCase\WPTestCase {
	private function make_instance() {
		return new Template_Bootstrap();
	}

> 	/**

Reviewer: For the sake of example, this test method can be removed.

	 * @test
	 */
	public function it_should_be_instantiatable() {
		$sut = $this->make_instance();

		$this->assertInstanceOf( Template_Bootstrap::class, $sut );
	}

	public function base_template_options() {
		return [
			'invalid' => [
				'foo',
				'event',
			],
			'numeric' => [
				2,
				'event',
			],
			'default' => [
				'default',
				'page',
			],
			'empty_string' => [
				'',
				'event',
			],
			'numeric_zero' => [
				0,
				'event',
			],
			'null' => [
				null,
				'event',
			],
			'boolean_false' => [
				false,
				'event',
			],
			'boolean_true' => [
				false,
				'event',
			],
			'slug_event' => [
				'event',
				'event',
			],
			'slug_page' => [
				'page',
				'event',
			],
		];
	}

	/**
	 * @test
	 * @dataProvider base_template_options
	 */
	public function should_only_allow_permitted_values_on_base_template_option( $input, $expected ) {

> 		tribe_update_option( 'tribeEventsTemplate', $input );

Reviewer: ```suggestion
		// The bootstrapping depends on this option, rather than arguments, here we "mock" the test context.
		tribe_update_option( 'tribeEventsTemplate', $input );
```

> 		$option_value = $this->make_instance()->get_template_setting();

Reviewer: ```suggestion
		$template_bootstrap = new Template_Bootstrap();
		$template_setting = $template_bootstrap->get_template_setting();
```

> 		$this->assertEquals( $option_value, $expected );

Reviewer: ```suggestion
		// PHPUnit will format a message for us comparing the two strings.
		$this->assertEquals( $template_setting, $expected );
```

	}

	// ...
}
```

> There are quite a few examples of things that would benefit from tests of that nature, a good separate example is related to how we have multiple display settings for formatting date and time depending on a number of coditionals, a good behavior is to create [snapshot testing](#snapshot-testing) to make sure those criteria are met when the setting is modified.

Reviewer: ```suggestion
There are many examples of things that would benefit from tests like this one.
A good example is the handling of multiple display settings to format date and time depending on a number of arguments and settings: creating [snapshot tests](#snapshot-testing) to cover all the combinations would be another example of this "accessory" testing.
```

> Remembering that the intention of these tests is related to not only testing positive behavior but also negative, which has to be met with a certain level of questioning to prevent over-engineering of those.

Reviewer: ```suggestion
These tests should test positive and negative behavior alike to ensure correct handling of "good" and "bad" combinations of arguments and settings.
```
